### DISCUSSION TITLE
Micro:bit v2 support

### DISCUSSION THREAD

**Discussion by heharkon** [pyocd/pyOCD#1161] [General]
Hi!

I purchased couple of BBC Micro:bits and seems that there's a new version, v2 of it, which I got then. It uses nrf52 instead of nrf51, so the existing board info for micro:bit doesn't match, I get 

`Board ID 9904 is not recognised, using generic cortex_m target.` and then fail... 

9900 and 9901 seem to be in the list for the first generation of Micro:bit.

I'm humbly asking would it be possible to support this version also? And if there's something I could help with, there might be a possibility to help. 😃 I do python, but I'm a total beginner with the boards etc. 😄

**Comment by flit** [comment]
Sorry… I have a partially implemented change but haven't had time to finish it. It's on my list.

Here's the branch: https://github.com/flit/pyOCD/tree/feature/microbit_v2
This has the board ID and a test binary.

The main thing missing is a built-in target for the nRF52833. If you'd like to help, you could duplicate the [built-in nRF52840 target support](https://github.com/pyocd/pyOCD/blob/master/pyocd/target/builtin/target_nRF52840_xxAA.py) and correct the memory map.

There is some general guidance for adding a new built-in target in [adding_new_targets.md](https://github.com/pyocd/pyOCD/blob/master/docs/adding_new_targets.md).

**Reply by heharkon** [reply]
Ok, and thanks for the quick reply! 

Great to hear that it's cooking! I can take a look at that memory map, just not sure if I understand any of it, but otoh, I'm here to learn, so... 😄

**Reply by flit** [reply]
I'm happy to help with any questions you have.

**Comment by flit** [comment]
@heharkon Fyi, I added micro:bit v2 and builtin nRF52833 support on this branch of my fork: [feature/microbit_v2](https://github.com/flit/pyOCD/tree/feature/microbit_v2).

If you have a chance, would you mind testing it?

You can install directly from this branch: `pip install -U --pre git+https://github.com/flit/pyOCD.git@feature/microbit_v2`

Thanks!

**Reply by heharkon** [reply]
Thank you @flit !

I haven't had too much time at the computer during the beginning of summer, but I tested this briefly.

It seems that there comes a version conflict:

`pyocd-pemicro 1.0.4 requires pyocd>=0.30.0, but you'll have pyocd 0.28.1.dev294 which is incompatible.`

and running pyocd halts with a stacktrace. Should I pick some older version of pyocd-pemicro?

**Reply by flit** [reply]
Thanks for trying it, sorry you had trouble!

You might have to uninstall pyocd then reinstall.

That error means the version of the installed pyocd didn't get updated for some reason from your previously installed version. The version should be something like `0.30.4.dev57`. I tested it in a fresh virtualenv just now and it installed ok.

**Reply by heharkon** [reply]
Ah, new virtualenv helped.

Now it seems to find my device when it's plugged in, yay!

```
pyocd list
  #   Probe                     Unique ID
----------------------------------------------------------------------------------
  0   micro:bit v2 [nrf52833]   9904360258994e45005e900f00000055000000009796990b
```

Thanks!

**Reply by flit** [reply]
Great! Have you been able to test flashing and debugging? Thanks.

**Reply by heharkon** [reply]
I tried now with an Ada project, but it failed. One issue here is, that also the Ada Drivers Library support for the v2 is preliminary, so I'm not sure if this error is because of the compilation of the binary or pyocd flashing. Does this error message say anything to you? I can report back to the driver's library if something is wrong there. 😄 

```
gprbuild --target=arm-eabi -d -PC:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\text_scrolling.gpr C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\src\main.adb -largs -Wl,-Map=map.txt
gprbuild: "main" up to date
[2021-06-29 21:29:34] process terminated successfully, elapsed time: 05.40s
Retrieving the load address.
arm-eabi-objdump C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\obj\main -h
Load address is: 0x00000000
Creating the binary (flashable) image.
arm-eabi-objcopy -O binary C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\obj\main C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\obj\main.bin
Flashing image to board...
pyocd flash -a 0x00000000 C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\obj\main.bin
[---|---|---|---|---|---|---|---|---|----]
[0001718:CRITICAL:__main__:('erase_sector(0x0) error: 103', 0, 103)
Traceback (most recent call last):
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\__main__.py", line 150, in run
    status = cmd.invoke()
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\subcommands\flash_cmd.py", line 90, in invoke
    programmer.program(self._args.file,
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\flash\file_programmer.py", line 156, in program
    self._loader.commit()
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\flash\loader.py", line 164, in commit
    perf = builder.program(chip_erase=chipErase,
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\flash\builder.py", line 470, in program
    flash_operation = self._sector_erase_program_double_buffer(progress_cb)
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\flash\builder.py", line 867, in _sector_erase_program_double_buffer
    self.flash.erase_sector(sector.addr)
  File "c:\users\hehar\koe\venv\lib\site-packages\pyocd\flash\flash.py", line 367, in erase_sector
    raise FlashEraseFailure('erase_sector(0x%x) error: %i' % (address, result), address, result)
pyocd.core.exceptions.FlashEraseFailure: ('erase_sector(0x0) error: 103', 0, 103)
Could not flash the executable.
[workflow stopped]
```

**Reply by flit** [reply]
It looks like it's something with the flash algo reporting an error, rather than the Driver Library code. Could you send me the ELF (`C:\Users\hehar\dev\Ada_Drivers_Library\examples\MicroBit_v2\text_scrolling\obj\main`) so I can use it for testing? (Either attach here or send to me via email.) Thanks for helping out! And sorry it's not working straight away.

**Reply by heharkon** [reply]
Sure, here it is, zipped. No worries, I'm glad that you have time to work with this! 

[main.zip](https://github.com/pyocd/pyOCD/files/6738762/main.zip)

**Reply by flit** [reply]
Thanks!

**Reply by flit** [reply]
It looks like the failure is related to the Nordic SoftDevice and its protection.

First, I'm able to reproduce the failure you see with the firmware you attached above using `pyocd flash -u9904 main.elf` (I added the .elf extension myself).

Yet, I can successfully first perform a chip erase (`pyocd erase -u9904 --chip`) and then program the test binary at `./test/data/binaries/microbitv2.bin` with default settings: `pyocd flash -u9904 main.elf`. However, the flash command after chip erase reports 34 pages skipped—this only happens when the data already in flash is the same as that being programmed. That should never be the case just after a chip erase (unless you are programming blank data, which is not the case here, double checked by examining the test file).

Those 34 pages that weren't erased almost certainly contain the SoftDevice.

When I either force pyocd to use sector erase, or use chip erase when I turn off the `smart_flash` option that controls the skipping of programming the same data, it fails as expected.

```
► pyocd flash -u9904 test/data/binaries/microbitv2.bin -esector -Osmart_flash=0 -Okeep_unwritten=0 -Odebug.traceback=0
[                    ]   1%0013955:CRITICAL:__main__:('erase_sector(0x0) error: 103', 0, 103)
```

```
► pyocd flash -u9904 test/data/binaries/microbitv2.bin -echip -Osmart_flash=0 -Okeep_unwritten=0 -Odebug.traceback=0
[                    ]   1%0013955:CRITICAL:__main__:('program_page(0x0) error: 103', 0, 103)
```

Right now I don't know how to work around it. The answer will require some investigation and reading of the nRF52 reference manual. It might require a mass erase using the nRF52's CTRL-AP, which will require some additions to pyocd (fortunately that I've already started).

**Reply by flit** [reply]
Umm, actually… I completely forgot that nRF52 mass erase is already supported in pyocd!

I was able to successfully program your test image by performing a mass erase (`pyocd erase --mass`) to wipe the SoftDevice, then programming as usual.

```
► pyocd erase -u9904 --mass
0005466:INFO:eraser:Mass erasing device...
0005676:INFO:eraser:Successfully erased.
► pyocd flash -u9904 var/devs/nrf52833/main.elf 
[====================] 100%
0005944:INFO:loader:Erased 4096 bytes (1 sector), programmed 4096 bytes (1 page), skipped 0 bytes (0 pages) at 10.44 kB/s
```

Once you remove the SoftDevice with the mass erase, you can program flash normally, until and unless you program an image that contains a SoftDevice.

**Reply by heharkon** [reply]
That worked, thank you very much!! 😃

**Reply by flit** [reply]
Glad to hear it. Thanks for helping me test!

**Comment by flit** [comment]
fyi: #1177

**Comment by flit** [comment]
micro:bit v2 support is integrated into version [0.31.0](/pyocd/pyOCD/releases/v0.31.0) of pyocd.
